parsl/dataflow/memoization.py

```diff
@@ -69,10 +146,22 @@ def make_hash(self, task):
             - hash (str) : A unique hash string
         """
         # Function name TODO: Add fn body later
-        t = [serialize_object(task['func_name'])[0],
-             serialize_object(task['fn_hash'])[0],
-             serialize_object(task['args'])[0],
-             serialize_object(task['kwargs'])[0]]
+
+        # if kwargs contains an outputs parameter, that parameter is removed
+        # and normalised differently - with output_ref set to True.
+        if 'outputs' in task['kwargs']:
```

Comment:
Inputs exist before invocation, at memoization time, in the same way that a normal arg or kwargs does, so id_for_memo can fully examine the input if it wants to (for example, I have prototyped taking a checksum of an input file so that checkpoints are invalidated if the contents of an input file changes - see the feature request at https://github.com/Parsl/parsl/issues/105).

In contrast to normal args, kwargs, and `inputs`, an `output` does not have content at memoization time. All that exists is the *name* of the output. (although there might be a file at the indicated name already, that will be overwritten and so is irrelevant)

So an id_for_memo which is more aware of #105-style content needs to know if it is looking at an input (in which case it can look at the content of the input) or an output (in which case it can only look at the name of the output).

Reply:
Okay, makes sense.